### PR TITLE
ndt2 changes

### PR DESCRIPTION
**Pull request by AlexandreAndr**

### CODE REVIEW

**Review by mazabou — COMMENTED** [review]
First batch of comments. Outside of the shuffle on the batch, looks good overall! just small suggestions

File: torch_brain/models/ndt2.py [hunk 1]
```diff
@@ -0,0 +1,682 @@
...
+        # if self.unsorted:
+        #     chan_nb_mapper = self.extract_chan_nb(data.units)
+        #     spikes.unit_index = chan_nb_mapper.take(spikes.unit_index)
+        #     # TODO do not work need to find an hack
+        #     # nb_units = chan_nb_mapper.max() + 1
+        #     num_units = 96
+
+        units_bincount = bin_spikes(
+            data.spikes, num_units, self.bin_time, dtype=np.int32
+        )
```
[the remark this answers is not in the document]
Comment (AlexandreAndr) [reply torch_brain/models/ndt2.py:204]: I think it is still right=True

File: torch_brain/models/ndt2.py [hunk 2]
```diff
@@ -0,0 +1,661 @@
...
+    def forward(
+        self,
+        latents: TensorType["b", "n_in + n_lat", "dim", int],
+        time_idx: TensorType["b", "n_in + n_lat", int],
+        space_idx: TensorType["b", "n_in + n_lat", int],
+        latent_mask: TensorType["b", "n_in + n_lat", bool],
+        decoder_attn_mask: TensorType[
+            "b x decoder_heads", "n_in + n_lat", "n_in + n_lat", bool
+        ],
+        ctx_emb: Optional[TensorType["b", "n_ctx", "dim", float]],
```
Comment (mazabou) [inline torch_brain/models/ndt2.py:538]: TODO check if they are detach for SSL or just bhvr

File: torch_brain/models/ndt2.py [hunk 3]
```diff
@@ -0,0 +1,661 @@
...
+        )
+
+        decoder_attn_mask = latent_time_idx[:, :, None] < latent_time_idx[:, None, :]
+        decoder_attn_mask = F.pad(
+            decoder_attn_mask, (0, 0, 0, nb_ctx_tokens), value=True
+        )
+        decoder_attn_mask = F.pad(decoder_attn_mask, (0, nb_ctx_tokens), value=False)
+        decoder_attn_mask = repeat(
+            decoder_attn_mask, "b t1 t2-> (b h) t1 t2", h=decoder_heads
+        )
```
[the remark this answers is not in the document]
Comment (AlexandreAndr) [reply torch_brain/models/ndt2.py:431]: cannot been removed 😢

**Review by mazabou — COMMENTED** [review]
This looks really solid, thanks Alex! Did you try to visualize the attention mask for the encoder then the decoder to verify that you are getting the expected masking?

File: torch_brain/models/ndt2.py [hunk 4]
```diff
@@ -0,0 +1,921 @@
...
+        self.bin_size = int(np.round(ctx_time / bin_time))
+        self.max_bincount = max_bincount
+        self.units_per_patch = units_per_patch
+        self.max_time_patches = max_time_patches
+        self.max_space_patches = max_space_patches
+        self.mask_ratio = mask_ratio
+
+        # TODO Check NDT2 init_scale
+        self.patch_emb = Embedding(
+            max_bincount + 1, dim // units_per_patch, padding_idx=max_bincount
```
[the remark this answers is not in the document]
Comment (AlexandreAndr) [reply torch_brain/models/ndt2.py:58]: Yes it is, I'll do the error raising. Are we keeping this design? 
I find it pretty consistent but we can also try other design (e.g., having the ending for every patch of dim side and pooling them to obtain the dim embedding).
What is your insight on it?

File: torch_brain/models/ndt2.py [hunk 5]
```diff
@@ -0,0 +1,921 @@
...
+            "target": pad(tgt_units_patch),
+            "extra_units_mask": pad(extra_units_mask),
+        }
+
+        return {
+            "context_tokens": context_tokens,
+            "encoder_tokens": encoder_tokens,
+            "decoder_tokens": decoder_tokens,
+            "target_tokens": target_tokens,
+        }
```
[the remark this answers is not in the document]
Comment (AlexandreAndr) [reply torch_brain/models/ndt2.py:263]: I refactored it tell me if it correspond more to what we want as standard :)!

**Comment by AlexandreAndr** [comment]
> This looks really solid, thanks Alex! Did you try to visualize the attention mask for the encoder then the decoder to verify that you are getting the expected masking?

The SSL training is working and seems consistent (SSL loss with IBL data are in reasonable range in [0.2, 0.3] after few epoch). 
I can also do via if you think it is important!
